### PR TITLE
Simulated Annealing for ANN_MLP training method

### PR DESCRIPTION
**Pull request by LaurentBerger**
**Merge with extra**: https://github.com/opencv/opencv_extra/pull/412

Add a trainning method to ANN_MLP

```
opencv_extra=metrop2
```

### CODE REVIEW

**Comment by alalek** [comment]
@LaurentBerger Please use the same branch name for opencv/opencv_extra PRs.

File: modules/ml/test/test_mltests2.cpp [hunk 1]
```diff
@@ -241,10 +243,62 @@ TEST(ML_ANN, ActivationFunction)
...
+        x->save(dataname + methodName[i] + ".yml");
+#else
+        Ptr<ml::ANN_MLP> y = Algorithm::load<ANN_MLP>(dataname + methodName[i] + ".yml");
+        ASSERT_TRUE(y != NULL) << "Could not load   " << dataname + methodName[i] + ".yml";
+        Mat testSamples = tdata->getTestSamples();
+        Mat rx, ry, dst;
+        x->predict(testSamples, rx);
+        y->predict(testSamples, ry);
+        double n = cvtest::norm(rx, ry, NORM_INF);
+        ASSERT_TRUE(n<FLT_EPSILON) << "Predict are not equal for " << dataname + methodName[i] + ".yml and " << methodName[i];
```
Comment (alalek) [inline modules/ml/test/test_mltests2.cpp]: Consider to use
```
EXPECT_LT(n, FLT_EPSILON) << ...
```
here and above for better error messages.

**Comment by LaurentBerger** [comment]
About ["Expected: (n) < (1.19209289550781250000e-7F), actual: 3.4318 vs 1.19209e-07" ](http://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/13686/steps/test_ml/logs/tests%20summary) in [saved file](https://github.com/LaurentBerger/opencv_extra/blob/97347e4555b0d4224130b6dbf06b54d65d26e64f/testdata/ml/waveform_backprop.yml#L16) moment_scale: 0.. I try to set on my windows system moment_scale:  0.1 and I have got exactly same error (reading 0 on linux gives -0 and [hence ](https://github.com/opencv/opencv/blob/master/modules/ml/src/ann_mlp.cpp#L125-L126
)
reading 0 on linux gives -0 ? Is it right?

**Comment by LaurentBerger** [comment]
@alalek  I haven't got any idea what's problem origin. yml files for RPROP and ANNEAL are ok for all platforms but BACKPROP is wrong for linux and macos plaform (but good for windows platform)

**Comment by alalek** [comment]
1. RNG for float/double numbers is not bit-exact between platforms. Use RNG for integers: `rng.uniform(0.0, 1.0)` ==> `rng.uniform(0, 1<<N) * (1.0 / (1 << N));` (N=31 or 16, division by 2^n is bitexact)
2. RNG fields must be removed - to get reproducible results use cv::theRNG() directly or initialize local RNG with seed from `theRNG()`.
3. Consider to use PImpl idiom for public interfaces to avoid virtual methods / exposed filelds (see cv::Mutex / cv::ocl::Context / etc).

**Comment by LaurentBerger** [comment]
About RNG I use RNG class in ANNEAL with int and float. I don't understand why backprop test is false and not anneal.

Now I have found a linux machine and I can reproduce error with backprop.  
I changed my program : I generate weight for all algorithms using 10 iteration of RPROP, save weights and use this weights for initial condition (ml::ANN_MLP::UPDATE_WEIGHTS).
RPROP and ANNEAL test are true and  backprop false.
I disable suffle data in backprop: 
```
                // shuffle indices
                for( i = 0; i <0*count; i++ ) // DISABLE SHUFFLE
                {
                    j = rng.uniform(0, count);
                    k = rng.uniform(0, count);
                    std::swap(_idx[j], _idx[k]);
                }

```
 backprop test : false

I change max-iter value:

`        int iter = -1, max_iter = termCrit.maxCount;//termCrit.maxCount*count;
`

and now test is : 

```
    for (size_t i = 0; i < methodType.size(); i++)
    {
        rng.state = 1027401484159173092;
        Ptr<ml::ANN_MLP> x = Algorithm::load<ANN_MLP>(dataname + "_init_weight.yml");
//        Ptr<ml::ANN_MLP> x = ml::ANN_MLP::create();
        x->setTrainMethod(methodType[i]);
        x->setTermCriteria(TermCriteria(TermCriteria::COUNT, 2, 0.01));
        for (int jj = 0; jj<10; jj++)
        {
            x->train(tdata, ml::ANN_MLP::NO_OUTPUT_SCALE + ml::ANN_MLP::UPDATE_WEIGHTS);
            x->save(format("%s%s%d.yml", dataname.c_str(), methodName[i].c_str(), jj));

        }
        ASSERT_TRUE(x->isTrained()) << "Could not train networks with  " << methodName[i];

```
and I compare weights values saved and I can see a small shift in data after four iteration, linux file : 
```
        - 9.4769785869628331e-01
         - -9.7244867343035279e-01
         - -1.0317135702854217e+00
      -
         - -7.8304381284119684e-01
         - 6.0925374767425977e-01
         - 1.1084102001850689e+00

```
windows files

```
         - 9.4769785869628331e-01
         - -9.7244867343035279e-01
         - -1.0317135702854217e+00
      -
         - -7.8304381284119673e-01 <--
         - 6.0925374767425977e-01
         - 1.1084102001850689e+00

```

That's only four iterations and real test it 4000 (training samples)*10=40000 iterations.

**Comment by alalek** [comment]
Could you try to save in YML+BASE64 format and check saved files?

**Comment by LaurentBerger** [comment]
My program is here. I change size of train sample size and layer number to get a progressive results (only 50 data to train a Neural Network  with 3 layers and 30 neurons)

first weight in NN for windows 
ww=[8.7918533220495676e-01 8.8378553813675320e-01  9.1630792217943446e-01 9.3592606046773896e-01 9.4115429560684061e-01 9.1629047458588930e-01 9.1344253498599737e-01 9.3428212428097301e-01 9.1699321878620421e-01 9.0443458316643710e-01]
for linux 
wl=[8.7918533220495676e-01 8.8378553813673577e-01 9.1630792218182699e-01 9.3592606045661697e-01 9.4115429557002528e-01 9.1629047449564982e-01 9.1344253536938202e-01  9.3428211832646124e-01 9.1699322891487989e-01 9.0443384297070883e-01];

Curve in log (abs(wl-ww)) is
image

source code is [here](https://gist.github.com/LaurentBerger/dbcd31f253c3bec7f842e892e4f84576#file-googletest-L773-L819)
[waveform_backprop_windows.zip](https://github.com/opencv/opencv/files/1535332/waveform_backprop_windows.zip)
[waveform__backprop_linux.zip](https://github.com/opencv/opencv/files/1535333/waveform__backprop_linux.zip)

File: modules/ml/include/opencv2/ml.hpp [hunk 2]
```diff
@@ -1499,6 +1502,34 @@ class CV_EXPORTS_W ANN_MLP : public StatModel
     /** @copybrief getRpropDWMax @see getRpropDWMax */
     CV_WRAP virtual void setRpropDWMax(double val) = 0;
 
+    /** ANNEAL: Update initial temperature.
+    It must be \>=0. Default value is 10.*/
+    /** @see setAnnealInitialT */
+    CV_WRAP virtual double getAnnealInitialT() const = 0;
```
Comment (alalek) [inline modules/ml/include/opencv2/ml.hpp]: There is no way to alter virtual function table without breaking of library ABI interface.

Try using this:
```
// .hpp file
class CV_EXPORTS_W ANN_MLP : public StatModel
{
    ...
    CV_WRAP double getAnnealInitialT() const = 0; // <-- no virtual;
    ... other new methods ...
}

class CV_EXPORTS_W ANN_MLP_ANNEAL : public ANN_MLP
{
    ... new virtual methods ...
}

//ann_mlp.cpp

double ANN_MLP::getAnnealInitialT() const
{
    const ANN_MLP_ANNEAL* this_ = dynamic_cast<const ANN_MLP_ANNEAL*>(this);
    if (!this_)
        CV_Error(Error::StsNotImplemented, "the class is not ANN_MLP_ANNEAL");
    return this_->getAnnealInitialT();
}

//replace class ANN_MLPImpl : public ANN_MLP
class ANN_MLPImpl : public ANN_MLP_ANNEAL
```

Also there is option without "ANN_MLP_ANNEAL" class, but "dynamic_cast" is used with "ANN_MLPImpl" directly.

Comment (LaurentBerger) [reply modules/ml/include/opencv2/ml.hpp]: I try and I must copy all methods in ANN_MLPImpl isn't it?

Comment (alalek) [reply modules/ml/include/opencv2/ml.hpp]: All **new** methods.

Comment (LaurentBerger) [reply modules/ml/include/opencv2/ml.hpp]: In ann_mlp.cpp I insert: 
class ANN_MLP_ANNEALImpl : public ANN_MLP_ANNEAL
{
public:
    ANN_MLP_ANNEALImpl()
    {
        clear();
        setActivationFunction(SIGMOID_SYM, 0, 0);
        setLayerSizes(Mat());
        setTrainMethod(ANN_MLP::ANNEAL, 0.1, FLT_EPSILON, 0, 0);
    }
    virtual ~ANN_MLP_ANNEALImpl() {}
        CV_IMPL_PROPERTY(double, AnnealInitialT, params.initialT)
        CV_IMPL_PROPERTY(double, AnnealFinalT, params.finalT)
        CV_IMPL_PROPERTY(double, AnnealCoolingRatio, params.coolingRatio)
        CV_IMPL_PROPERTY(int, AnnealItePerStep, params.itePerStep)
        
     AnnParams params;
};

Ptr<ANN_MLP_ANNEAL> ANN_MLP_ANNEAL::create()
{
    return makePtr<ANN_MLP_ANNEALImpl>(); <---- ERROR  2259
}


>G:\Lib\opencv\modules\core\include\opencv2/core/ptr.inl.hpp(301): error C2259: 'cv::ml::ANN_MLP_ANNEALImpl' : class' : cannot instantiate abstract class
see  G:\Lib\opencv\modules\core\include\opencv2/core/ptr.inl.hpp(301): note: 'int cv::ml::StatModel::getVarCount(void) const' :  is abstract

and all methods are enumerated


My anneal class is
class CV_EXPORTS_W ANN_MLP_ANNEAL : public ANN_MLP
{
public:
    /** ANNEAL: Update initial temperature.
    It must be \>=0. Default value is 10.*/
    /** @see setAnnealInitialT */
    CV_WRAP virtual double getAnnealInitialT() const;
    /** @copybrief getAnnealInitialT @see getAnnealInitialT */
    CV_WRAP virtual void setAnnealInitialT(double val) const;

    /** ANNEAL: Update final temperature.
    It must be \>=0 and less than initialT. Default value is 0.1.*/
    /** @see setAnnealFinalT */
    CV_WRAP  virtual double getAnnealFinalT() const;
    /** @copybrief getAnnealFinalT @see getAnnealFinalT */
    CV_WRAP  virtual void setAnnealFinalT(double val) ;

    /** ANNEAL: Update cooling ratio.
    It must be \>0 and less than 1. Default value is 0.95.*/
    /** @see setAnnealCoolingRatio */
    CV_WRAP  virtual double getAnnealCoolingRatio() const ;
    /** @copybrief getAnnealCoolingRatio @see getAnnealCoolingRatio */
    CV_WRAP  virtual void setAnnealCoolingRatio(double val) const;

    /** ANNEAL: Update iteration per step.
    It must be \>0 . Default value is 10.*/
    /** @see setAnnealItePerStep */
    CV_WRAP virtual int getAnnealItePerStep() const;
    /** @copybrief getAnnealItePerStep @see getAnnealItePerStep */
    CV_WRAP virtual  void setAnnealItePerStep(int val) const;


    /** @brief Creates empty model

    Use StatModel::train to train the model, Algorithm::load\<ANN_MLP\>(filename) to load the pre-trained model.
    Note that the train method has optional flags: ANN_MLP::TrainFlags.
    */
    CV_WRAP static Ptr<ANN_MLP_ANNEAL> create();

    /** @brief Loads and creates a serialized ANN from a file
    *
    * Use ANN::save to serialize and store an ANN to disk.
    * Load the ANN from this file again, by calling this function with the path to the file.
    *
    * @param filepath path to serialized ANN
    */
    CV_WRAP static Ptr<ANN_MLP_ANNEAL> load(const String& filepath);
....

**Comment by berak** [comment]
@alalek , just curious, would it be possible, to keep the original design (extending the ANN_MLP interface with some virtual methods), and just wait with the merge until 3.4 (which is like, around the corner) ?

like: https://github.com/opencv/opencv/labels/future

File: modules/ml/include/opencv2/ml.hpp (hunk 2 above)
Comment (alalek) [reply modules/ml/include/opencv2/ml.hpp]: > In ann_mlp.cpp I insert

Need just replace base class `ANN_MLP` => `ANN_MLP_ANNEAL` of `ANN_MLPImpl` (not `ANN_MLP_ANNEALImpl`).

**Comment by alalek** [comment]
"future" means OpenCV 4.0

"PImpl" idiom allows to add new methods easy - there are no "virtual" methods at all.

**Comment by LaurentBerger** [comment]
@alalek Something is wrong in my PR. I have got this error : 
[C++ exception with description "/build/precommit_linux64/opencv/modules/ml/src/ann_mlp.cpp:161: error: (-213) the class is not ANN_MLP_ANNEAL in function setAnnealInitialT](http://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/13829/steps/test_ml/logs/stdio)
Error is [here ](https://github.com/LaurentBerger/opencv/blob/metrop2/modules/ml/test/test_mltests2.cpp#L308)

File: modules/ml/src/ann_mlp.cpp [hunk 3]
```diff
@@ -72,13 +83,140 @@ inline T inBounds(T val, T min_val, T max_val)
...
+
+void ANN_MLP::setAnnealItePerStep(int val) const
+{
+    const ANN_MLP_ANNEAL* this_ = dynamic_cast<const ANN_MLP_ANNEAL*>(this);
+    if (!this_)
+        CV_Error(Error::StsNotImplemented, "the class is not ANN_MLP_ANNEAL");
+    this_->setAnnealItePerStep(val);
+}
+
 class ANN_MLPImpl : public ANN_MLP
```
Comment (alalek) [inline modules/ml/src/ann_mlp.cpp]:
```.diff
-class ANN_MLPImpl : public ANN_MLP
+class ANN_MLPImpl : public ANN_MLP_ANNEAL
```

File: modules/ml/include/opencv2/ml.hpp [hunk 4]
```diff
@@ -1838,6 +1867,166 @@ CV_EXPORTS void randMVNormal( InputArray mean, InputArray cov, int nsamples, Out
...
+    /** @copybrief getAnnealItePerStep @see getAnnealItePerStep */
+    CV_WRAP virtual  void setAnnealItePerStep(int val) const = 0;
+
+
+    /** @brief Creates empty model
+
+    Use StatModel::train to train the model, Algorithm::load\<ANN_MLP\>(filename) to load the pre-trained model.
+    Note that the train method has optional flags: ANN_MLP::TrainFlags.
+    */
+    CV_WRAP static Ptr<ANN_MLP> create();
```
Comment (alalek) [inline modules/ml/include/opencv2/ml.hpp]: Remove this `create()`. Reuse existed from ANN_MLP.

File: modules/ml/include/opencv2/ml.hpp [hunk 5]
```diff
@@ -1838,6 +1867,166 @@ CV_EXPORTS void randMVNormal( InputArray mean, InputArray cov, int nsamples, Out
...
+        initialT = 2;
+        finalT = 0.1;
+        coolingRatio = 0.95;
+        iterPerStep = 100;
+    }
+    RNG rEnergy;
+    double coolingRatio;
+    double initialT;
+    double finalT;
+    int iterPerStep;
```
Comment (alalek) [inline modules/ml/include/opencv2/ml.hpp]: Implementation details should be moved out from public headers.
All details should be stored in `SimulatedAnnealingSolver::Impl`, which should be placed in .cpp file.

File: modules/ml/include/opencv2/ml.hpp [hunk 6]
```diff
@@ -1838,6 +1867,166 @@ CV_EXPORTS void randMVNormal( InputArray mean, InputArray cov, int nsamples, Out
...
+    Impl* impl;
+};
+struct SimulatedAnnealingSolver::Impl
```
Comment (alalek) [inline modules/ml/include/opencv2/ml.hpp]: see comment above

File: modules/ml/src/ann_mlp.cpp [hunk 7]
```diff
@@ -42,6 +42,10 @@
 
 namespace cv { namespace ml {
 
+    double SimulatedAnnealingSolver::energy() { return impl->energy(); };
```
Comment (alalek) [inline modules/ml/src/ann_mlp.cpp]: Please avoid Indentation in namespaces.

File: modules/ml/src/ann_mlp.cpp (hunk 3 above)
Comment (LaurentBerger) [reply modules/ml/src/ann_mlp.cpp]: class ANN_MLPImpl : public ANN_MLP_ANNEAL
error C2259 cannot instantiate abstract class

File: modules/ml/include/opencv2/ml.hpp [hunk 8]
```diff
@@ -1838,6 +1867,166 @@ CV_EXPORTS void randMVNormal( InputArray mean, InputArray cov, int nsamples, Out
...
+
+@sa @ref ml_intro_ann
+*/
+class CV_EXPORTS_W ANN_MLP_ANNEAL : public ANN_MLP
+{
+public:
+    CV_WRAP virtual void setTrainMethod(int method, double param1 = 0, double param2 = 0, double param3 = 0, int param4 = 0) = 0;    /** ANNEAL: Update initial temperature.
+    It must be \>=0. Default value is 10.*/
+    /** @see setAnnealInitialT */
+    CV_WRAP virtual double getAnnealInitialT() const=0;
```
Comment (alalek) [inline modules/ml/include/opencv2/ml.hpp]: Because you should not use `=0` here.
You should provide implementations for these methods in .cpp file.

Comment (LaurentBerger) [reply modules/ml/include/opencv2/ml.hpp]: image

File: modules/ml/src/ann_mlp.cpp [hunk 9]
```diff
@@ -1390,23 +1648,106 @@ class ANN_MLPImpl : public ANN_MLP
 };
 
 
+
+class ANN_MLP_ANNEALImpl : public ANN_MLPImpl
```
Comment (alalek) [inline modules/ml/src/ann_mlp.cpp]: This is not need as separate class. Merge this contents into ANN_MLPImpl.
- revert "=0" change in .hpp file (but keep "const" fix).
- remove these new calls with dynamic_cast: `double ANN_MLP_ANNEAL::getAnnealInitialT() const`

**Comment by LaurentBerger** [comment]
@alalek I hope that I didn't kill your commit

**Comment by LaurentBerger** [comment]
How to rebase with a merge ? 

About test, BACKPROP test is disabled

**Comment by vpisarev** [comment]
:+1:
